### DISCUSSION TITLE
Type safe object construction with default values

### DISCUSSION THREAD

**Discussion by mwisnicki** [colinhacks/zod#1339] [Q&A] [answered]
How can I construct new object with default values in a type safe manner?

```ts
const Person = z.object({
  name: z.string(),
  age: z.number().optional(),
  friends: z.array(z.string()).default([])
})
type Person = z.infer<typeof Person>;
```

I know I can use parse but it accepts argument of any type
```ts
const p = Person.parse({}); // TODO I want it to fail because name is required
```

I'm looking for something that would only accept `{name: string, age?: number, friends?: string[]}`.
Note that this constructor type is different than `type Person` because fields with defaults are optional.

I guess a related question is how to obtain constructor type.

**Comment by scotttrinh** [comment, marked as answer]
This is similar to this request: https://github.com/colinhacks/zod/discussions/1335

In general, Zod expects that you're using it at some boundary that doesn't know the type at compile-time, thus `unknown`. So, network, database, user input, those kinds of things. If you're just looking to use it as a "constructor", you can make your own wrapper like:

```ts
type PersonInput = z.input<typeof Person>;
const makePerson = (input: PersonInput): Person => Person.parse(input);
```

Note that it actually still validates the input so your incurring a runtime cost that you might not need. The linked discussion above kind of covers this use case where you do not want validation or you have already narrowed your type in some way where you don't want to incur this cost (database schema, type-safe network protocol, etc).

**Reply by mwisnicki** [reply]
Thanks this is close enough!
